defrag/modules/docs.py

```diff
@@ -0,0 +1,163 @@
+from defrag.modules.helpers.services_manager import Controllers, ServiceTemplate, ServicesManager
+from defrag.modules.helpers.requests import Req
+from defrag.modules.helpers import Query, QueryResponse
+from lunr import lunr
+from lunr.index import Index
+from bs4 import BeautifulSoup
+from pydantic import BaseModel
+from concurrent.futures import ProcessPoolExecutor
+import asyncio
+import json
+import sys
+from typing import Any, Dict, List, Tuple
+
+__MOD_NAME__ = "docs"
+
+
+class LunrDoc(BaseModel):
+    number: str
+    contents: str
+    permalink: str
+    title: str
+
+
+indexes = {
+    "leap": {
+        "url": "https://doc.opensuse.org/documentation/leap/reference/single-html/book-reference/index.html",
+        "index": None,
+    },
+    "tumbleweed": {
+        "url": "https://raw.githubusercontent.com/openSUSE/openSUSE-docs-revamped-temp/gh-pages/search/search_index.json",
+        "index": None
+    }
+}
+
+
+async def get_data(source: str) -> Any:
+    async with Req(indexes[source]["url"], closeOnResponse=False) as result:
+        if source == "tumbleweed":
+            return await result.read()
+        return await result.text()
+
+
+def make_soup(data_str: str) -> List[LunrDoc]:
+    root = BeautifulSoup(data_str, 'html.parser')
+    chapters = root.find_all(class_="chapter")
+    res = []
+    for c in chapters:
+        titles = c.find_all(class_="titlepage")
+        contents = [c.get_text() for c in c.find_all("p")]
+        for i, t in enumerate(titles):
+            names = t.find_all(class_="name")[0].get_text()
+            permalink = t.find_all(class_="permalink")[0]["href"]
+            number = t.find_all(class_="number")[0].get_text().strip()
+            res.append(LunrDoc(number=number, title=names,
+                       permalink=permalink, contents="".join(contents[i])))
+    return res
+
+
+def make_leap_index(raw: str) -> Index:
+    docs: List[LunrDoc] = make_soup(raw)
+    listed: List[Dict[str, Any]] = [doc.dict() for doc in docs]
+    idx = lunr(ref="permalink", fields=("contents", "title"), documents=listed)
+    return idx
+
+
+def make_tumbleweed_index(raw: bytes) -> Index:
+    serialized: Dict[str, Any] = json.loads(raw)["index"]
+    return Index.load(serialized)
+
+
+def thin_prime(source: str, item: Dict[str, Any]) -> Dict[str, Any]:
+    built = {k: item[k] for k in item if k in ["score", "ref"]}
+    built["source"] = source
+    return built
+
+
+def sorted_on_score(items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    return sorted(items, key=lambda x: x["score"], reverse=True)
+
+
+def search_index(idx: Index, source: str, keywords: str) -> List[Dict[str, Any]]:
+    return list(map(lambda item: thin_prime(source, item), idx.search(keywords)))
+
+
+def set_global_index(source: str, idx: Index) -> None:
+    global indexes
+    indexes[source]["index"] = idx
+
+
+def ready_to_index(sources: List[str]) -> bool:
+    for s in sources:
+        if not indexes[s]["index"]:
```

Comment:
Right here it fails for me when I try to search something. I run `curl http://localhost:8000/docs/single/zypper/leap` and get the following in the console:

```
INFO:     127.0.0.1:56364 - "GET /docs/single/zypper/leap HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 371, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 59, in __call__
    return await self.app(scope, receive, send)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/routing.py", line 580, in __call__
    await route.handle(scope, receive, send)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/routing.py", line 241, in handle
    await self.app(scope, receive, send)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/starlette/routing.py", line 52, in app
    response = await func(request)
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/fastapi/routing.py", line 216, in app
    raw_response = await run_endpoint_function(
  File "/home/jens/src/defrag-api/ven/lib64/python3.8/site-packages/fastapi/routing.py", line 149, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/jens/src/defrag-api/defrag/modules/docs.py", line 138, in search_single_source_docs
    if not ready_to_index([source]):
  File "/home/jens/src/defrag-api/defrag/modules/docs.py", line 93, in ready_to_index
    if not indexes[s]["index"]:
KeyError: 'zypper'
```